### PR TITLE
prov/shm: refactor and simplify protocol code

### PR DESCRIPTION
**Pull request by aingerson**
When the shm provider was first released, it only had
three protocol (inline, inject, iov), making the protocol
selection more or less straightforward. However, as the
provider has developed, three new protocols have been added
(mmap, sar, ipc). The addition of these checks and protocols
has added a lot of complexity to the code, resulting in
very convoluted and hard to follow code with many duplicate
checks.

This patch simplifies all of the protocol code by
isolating each protocol into separate functions, accessible
through an array of function pointers by protocol enum.

These functions are used by the msg, tagged, and rma calls
which all take similar code paths. The atomic calls are
separated since they follow a slightly different pattern
because of the extra atomic/datatype fields.

This patch also makes the following two indirectly related
changes:
- smr_complete_tx now takes in 64bit FI flags instead
of the 16bit SMR flag, saving instructions to convert
from SMR to FI for the completion flags. Add 64bit op
flags to tx_entry to save them for response entry
completions
- Remove redundant atomic fetch/compare inline path and
force all compare and fetch atomics through the inject.
The old code was not using this path anyway because
SMR_RMA_REQ was always set for fetch and compares. Removing
this path simplifies atomic protocol selection as well.

Signed-off-by: aingerson <alexia.ingerson@intel.com>

### CODE REVIEW

File: prov/shm/src/smr_atomic.c [hunk 1]
```diff
@@ -357,19 +380,19 @@ static ssize_t smr_atomic_inject(struct fid_ep *ep_fid, const void *buf,
...
 	} else if (total_len <= SMR_INJECT_SIZE) {
-		tx_buf = smr_freestack_pop(smr_inject_pool(peer_smr));
-		smr_format_inject_atomic(cmd, FI_HMEM_SYSTEM, 0, &iov, 1, NULL,
-					 0, NULL, 0, peer_smr, tx_buf);
+		ret = smr_do_atomic_inject(ep, peer_smr, id, peer_id,
+				ofi_op_atomic, 0, FI_HMEM_SYSTEM, 0, datatype,
+				op, &iov, 1, NULL, 0, NULL, 0, total_len,
+				NULL, 0);
+		if (ret)
+			goto unlock_region;
```
Comment (j-xiong) [inline prov/shm/src/smr_atomic.c:393]: what happens if total_len > SMR_INJECT_SIZE?  There is 'assert(count <= SMR_INJECT_SIZE)' at the beginning of this function, but count is different from total_len.

File: prov/shm/src/smr_msg.c [hunk 2]
```diff
@@ -350,19 +294,18 @@ static ssize_t smr_generic_inject(struct fid_ep *ep_fid, const void *buf,
...
-	cmd = ofi_cirque_next(smr_cmd_queue(peer_smr));
-	smr_generic_format(cmd, peer_id, op, tag, data, op_flags);
+	if (len <= SMR_MSG_DATA_LEN)
+		ret = smr_proto_ops[smr_src_inline](ep, peer_smr, id, peer_id,
+				op, tag, data, op_flags, FI_HMEM_SYSTEM, 0,
+				&msg_iov, 1, len, NULL);
+	else
+		ret = smr_proto_ops[smr_src_inject](ep, peer_smr, id, peer_id,
+				op, tag, data, op_flags, FI_HMEM_SYSTEM, 0,
+				&msg_iov, 1, len, NULL);
```
Comment (j-xiong) [inline prov/shm/src/smr_msg.c:305]: Since all the arguments are the same, it would have less code duplication if proto is set by the if ... else ... statement and the function is called after that.

File: prov/shm/src/smr_rma.c [hunk 3]
```diff
@@ -404,33 +333,26 @@ ssize_t smr_generic_rma_inject(struct fid_ep *ep_fid, const void *buf,
...
-	ofi_cirque_commit(smr_cmd_queue(peer_smr));
-	peer_smr->cmd_cnt--;
+	if (len <= SMR_MSG_DATA_LEN)
+		ret = smr_proto_ops[smr_src_inline](ep, peer_smr, id, peer_id,
+				ofi_op_write, 0, data, flags, FI_HMEM_SYSTEM, 0,
+				&iov, 1, len, NULL);
+	else
+		ret = smr_proto_ops[smr_src_inject](ep, peer_smr, id, peer_id,
+				ofi_op_write, 0, data, flags, FI_HMEM_SYSTEM, 0,
+				&iov, 1, len, NULL);
```
Comment (j-xiong) [inline prov/shm/src/smr_rma.c:351]: Same here.

File: prov/shm/src/smr_atomic.c (hunk 1 above)
Comment (aingerson) [reply prov/shm/src/smr_atomic.c:392]: Ah that assert is definitely incorrect. Thanks for catching!

**Comment by aingerson** [comment]
Addressed comments and fixed uninitialized variable causing CI failures.

**Review by j-xiong — APPROVED** [review]
